recipes/qt-advanced-docking-system/all/conanfile.py

```diff
@@ -26,70 +33,79 @@ class QtADS(ConanFile):
+    def build_requirements(self):
+        if not can_run(self):
+            self.tool_requires("qt/<host_version>", options={"gui": False, "widgets": False})
 
-        self._cmake = CMake(self)
-        self._cmake.definitions["ADS_VERSION"] = self.version
-        self._cmake.definitions["BUILD_EXAMPLES"] = "OFF"
-        self._cmake.definitions["BUILD_STATIC"] = not self.options.shared
+    def source(self):
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
 
-        self._cmake.configure()
-        return self._cmake
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.cache_variables["ADS_VERSION"] = self.version
+        tc.variables["BUILD_EXAMPLES"] = "OFF"
+        tc.variables["BUILD_STATIC"] = not self.options.shared
+        tc.generate()
+        deps = CMakeDeps(self)
+        deps.generate()
 
     def _patch_sources(self):
-        apply_conandata_patches(self)
-
-        replace_in_file(self,
-            f"{self.source_folder}/{self._source_subfolder}/src/ads_globals.cpp",
-            "#include <qpa/qplatformnativeinterface.h>",
-            f"#include <{self._qt_version}/QtGui/qpa/qplatformnativeinterface.h>"
+        qt_version = self.dependencies["qt"].ref.version
+        replace_in_file(self, os.path.join(self.source_folder, "src", "ads_globals.cpp"),
+                        "#include <qpa/qplatformnativeinterface.h>",
+                        f"#include <{qt_version}/QtGui/qpa/qplatformnativeinterface.h>",
         )
 
     def build(self):
         self._patch_sources()
-        cmake = self._configure_cmake()
+        cmake = CMake(self)
+        cmake.configure()
         cmake.build()
 
     def package(self):
-        cmake = self._configure_cmake()
+        cmake = CMake(self)
         cmake.install()
-        self.copy("LICENSE", dst="licenses", src=self._source_subfolder)
+        copy(self, "LICENSE", self.source_folder, os.path.join(self.package_folder, "licenses"))
         rmdir(self, os.path.join(self.package_folder, "license"))
         rmdir(self, os.path.join(self.package_folder, "lib", "cmake"))
 
     def package_info(self):
+        if Version(self.version) >= 4:
+            name = f"qt{self._qt_major}advanceddocking"
+            self.cpp_info.includedirs.append(os.path.join("include", name))
+            lib_name = f"{name}d" if self.settings.build_type == "Debug" else name
+        else:
+            lib_name = "qtadvanceddocking"
+
         if self.options.shared:
-            self.cpp_info.libs = ["qtadvanceddocking"]
+            self.cpp_info.libs = [lib_name]
         else:
             self.cpp_info.defines.append("ADS_STATIC")
-            self.cpp_info.libs = ["qtadvanceddocking_static"]
+            self.cpp_info.libs = [f"{lib_name}_static"]
+
+        if is_msvc(self) and self._qt_major >= 6:
+            # Qt 6 requires C++17 and a valid __cplusplus value
+            self.cpp_info.cxxflags.append("/Zc:__cplusplus")
```

Comment:
```suggestion
        if is_msvc(self) and self._qt_major >= 6:
            # Qt 6 requires C++17 and a valid __cplusplus value
            self.cpp_info.cxxflags.append("/Zc:__cplusplus")
```

Not sure about this, it ensures the __cplusplus macro reflects the correct C++ standard version being used during compilation. But then you are exporting to the downstream. Is it the case that some MSVC versions do not fit a C++17 feature and are breaking this project?

Reply:
I think there was a check against the `__cplusplus` define value in the code somewhere, but I don't see one in the code anymore. I'll simply drop it for now.